### PR TITLE
(disallow|require)SpacesInFunctionDeclaration: check export default f…

### PR DESCRIPTION
**Pull request by hzoo**
…unction

WIP? If we don't want to do this now that's cool since it's before it's needed.

Basically `export default function() {}` is a `FunctionExpression` right now but when we update `esprima` it's going to be `FunctionDeclaration`. After this I think I can push `babel-jscs`?

So thus it would have no `node.id` and should be skipped in this rule.

I commented out some test's since they won't pass without updating `esprima` at the moment. Or we could just remove them.

For #1376

### CODE REVIEW

**Comment by mrjoelkemp** [comment]
So that I understand, we're saying that exporting default functions will be the only time that a nameless function declaration is proper syntax? Otherwise, it's a syntax error.

File: lib/rules/disallow-spaces-in-function-declaration.js [hunk 1]
```diff
@@ -75,6 +75,9 @@ module.exports.prototype = {
         var beforeOpeningCurlyBrace = this._beforeOpeningCurlyBrace;
 
         file.iterateNodesByType(['FunctionDeclaration'], function(node) {
+            if (!node.id) {
```
Comment (mrjoelkemp) [inline lib/rules/disallow-spaces-in-function-declaration.js]: Please move the description of this PR to an issue (the explanation of why this needs to happen) and then reference the issue number in a comment above this line. The reasoning as to why we need to do this should be captured in the code.

File: test/specs/rules/disallow-spaces-in-function-declaration.js [hunk 2]
```diff
@@ -40,4 +40,30 @@ describe('rules/disallow-spaces-in-function-declaration', function() {
             assert(checker.checkString('function abc (){}').isEmpty());
         });
     });
+
+    describe('es6', function() {
+        beforeEach(function() {
+            checker.configure({ esnext: true });
+        });
+
+        it('should not report missing space before round brace in export default function', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/disallow-spaces-in-function-declaration.js]: Maybe move to: 

```js
describe('before round brace', function() {
  beforeEach(function() {
    checker.configure({ disallowSpacesInFunctionDeclaration: { beforeOpeningRoundBrace: true } });
  });

  it('should not report missing space in export default function', ...

})
```

File: test/specs/rules/disallow-spaces-in-function-declaration.js [hunk 3]
```diff
@@ -40,4 +40,30 @@ describe('rules/disallow-spaces-in-function-declaration', function() {
...
+            checker.configure({ disallowSpacesInFunctionDeclaration: { beforeOpeningRoundBrace: true } });
+            assert(checker.checkString('export default function() {}').isEmpty());
+        });
+
+        it.skip('should report missing space before round brace in export default function', function() {
+            checker.configure({ disallowSpacesInFunctionDeclaration: { beforeOpeningRoundBrace: true } });
+            assert(checker.checkString('export default function(){}').getErrorCount() === 1);
+        });
+
+        it('should not report missing space before curly brace in export default function', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/disallow-spaces-in-function-declaration.js]: Same comment as above: move to its own describe and use a beforeEach for what's common

File: test/specs/rules/require-spaces-in-function-declaration.js [hunk 4]
```diff
@@ -40,4 +40,30 @@ describe('rules/require-spaces-in-function-declaration', function() {
             assert(checker.checkString('function abc() {}').isEmpty());
         });
     });
+
+    describe('es6', function() {
+        beforeEach(function() {
+            checker.configure({ esnext: true });
+        });
+
+        it.skip('should report missing space before round brace in export default function', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/require-spaces-in-function-declaration.js]: Same comment as in previous file. Move to describe with beforeEach

**Comment by markelog** [comment]
Okay, lets wait for @mrjoelkemp "LGTM"

File: lib/rules/disallow-spaces-in-function-declaration.js (hunk 1 above)
Comment (hzoo) [inline lib/rules/disallow-spaces-in-function-declaration.js]: Ok makes sense - made the issue.

File: test/specs/rules/disallow-spaces-in-function-declaration.js (hunk 2 above)
Comment (hzoo) [inline test/specs/rules/disallow-spaces-in-function-declaration.js]: Should I keep the skipped tests?

**Comment by mrjoelkemp** [comment]
LGMT

**Comment by zxqfox** [comment]
> Coverage decreased (-0.06%) to 99.23%

I'm aware of this. ;-(

**Comment by hzoo** [comment]
Well I can't write the tests for the added code since no code will run that
part of the code until we update esprima.

On Thursday, May 14, 2015, Alexej Yaroshevich <notifications@github.com>
wrote:

> Coverage decreased (-0.06%) to 99.23%
>
> I'm aware of this. ;-(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jscs-dev/node-jscs/pull/1369#issuecomment-102095468>.
>


-- 
-Henry Zhu
Georgia Institute of Technology
Industrial Engineering '14
Computer Science Minor

**Comment by zxqfox** [comment]
@hzoo Np, let's just left an issue for that.

**Comment by hzoo** [comment]
Sounds good! Yeah even the tests now aren't being tested correctly because of the same issue - but the missing tests would be checking for an error so I just left them out (instead of skipping).

**Comment by mrjoelkemp** [comment]
Merge or no?

**Comment by hzoo** [comment]
Can merge - we have an issue to add a test for coverage in #1377 for after esprima is updated to return `FunctionDeclaration` with a null id. Esprima issue is https://github.com/jquery/esprima/issues/311

**Comment by hzoo** [comment]
Actually looking at it now - I think I should add the same check to `disallowSpacesInFunction` and `disallowAnonymousFunctions` (and for require)
